go-controller/pkg/node/gateway_iptables.go

```diff
@@ -321,39 +321,50 @@ func recreateIPTRules(table, chain string, keepIPTRules []iptRule) {
 // only incoming traffic on that IP will be accepted for NodePort rules; otherwise incoming traffic on the NodePort
 // on all IPs will be accepted. If gatewayIP is "", then NodePort traffic will be DNAT'ed to the service port on
 // the service's ClusterIP. Otherwise, it will be DNAT'ed to the NodePort on the gatewayIP.
-func getGatewayIPTRules(service *kapi.Service, gatewayIP string, nodeIPs []*net.IPNet) []iptRule {
+func getGatewayIPTRules(service *kapi.Service, gatewayIPs []string, nodeIPs []*net.IPNet) []iptRule {
 	rules := make([]iptRule, 0)
-	for _, svcPort := range service.Spec.Ports {
-		if util.ServiceTypeHasNodePort(service) {
-			err := util.ValidatePort(svcPort.Protocol, svcPort.NodePort)
-			if err != nil {
-				klog.Errorf("Skipping service: %s, invalid service NodePort: %v", svcPort.Name, err)
-				continue
-			}
-			err = util.ValidatePort(svcPort.Protocol, svcPort.Port)
-			if err != nil {
-				klog.Errorf("Skipping service: %s, invalid service port %v", svcPort.Name, err)
-				continue
-			}
-			var targetIP string
-			if gatewayIP == "" {
-				targetIP = service.Spec.ClusterIP
-			} else {
+	clusterIPs := util.GetClusterIPs(service)
+	for _, clusterIP := range clusterIPs {
```

Comment:
i dont think this is right. We shouldn't loop on cluster IP to determine calling getNodePortIPTRules for gateway IP. Really we should keep the code like it was, then just handle looping through IP family  in here:
```
			if gatewayIP == "" {
				rules = append(rules, getNodePortIPTRules(svcPort, nodeIP, service.Spec.ClusterIP, svcPort.Port)...)
			} else {
				rules = append(rules, getNodePortIPTRules(svcPort, nodeIP, gatewayIP, svcPort.NodePort)...)
			}
```

Reply:
I don't follow your example. `service.Spec.ClusterIP` is not "dual-stack compliant", this is what this commit focuses on.

Reply:
Sorry my code example was not intended to be taken literal, I was just trying to highlight the place to move the for loop to. Since the function says cluster ip only matters if gatewayIP is empty, then can move that loop to the place in the code where gatewayIP is found to be empty.

Reply:
After having gone through this patch again, I think the code is correct. 

> Since the function says cluster ip only matters if gatewayIP is empty

The cluster IP does not _only_ matter if the gateway IP is empty. The cluster IP is needed for external IP services too (independently of the gateway IP), and moreover: in local gateway mode dual-stack; we can use the cluster IP to configure the node port service with the correct gateway IP version.

Reply:
yes you still need it for external IP, but you do not always need it for nodePort. I think you can restructure this code to accommodate for both, without running the node port append rule 3 times for the same gatewayIP (assuming gatewayIP is not empty). Quick hack:

```
func getGatewayIPTRules(service *kapi.Service, gatewayIP string, nodeIP *net.IPNet) []iptRule {
	rules := make([]iptRule, 0)
	for _, svcPort := range service.Spec.Ports {
		if util.ServiceTypeHasNodePort(service) {
			err := util.ValidatePort(svcPort.Protocol, svcPort.NodePort)
			if err != nil {
				klog.Errorf("Skipping service: %s, invalid service NodePort: %v", svcPort.Name, err)
				continue
			}
			err = util.ValidatePort(svcPort.Protocol, svcPort.Port)
			if err != nil {
				klog.Errorf("Skipping service: %s, invalid service port %v", svcPort.Name, err)
				continue
			}
			if gatewayIP == "" {
                                for _, clusterIP := range util.GetClusterIPs(service)  {
                                    // needs ip family checking
				    rules = append(rules, getNodePortIPTRules(svcPort, nodeIP, clusterIP, svcPort.Port)...)
                                 }
			} else {
			    rules = append(rules, getNodePortIPTRules(svcPort, nodeIP, gatewayIP, svcPort.NodePort)...)    
			}
		}
		for _, externalIP := range service.Spec.ExternalIPs {
			err := util.ValidatePort(svcPort.Protocol, svcPort.Port)
			if err != nil {
				klog.Errorf("Skipping service: %s, invalid service port %v", svcPort.Name, err)
				continue
			}
                        for _, clusterIP := range util.GetClusterIPs(service)  {
                            // needs ip family checking
			    rules = append(rules, getExternalIPTRules(svcPort, externalIP, clusterIP)...)
                        }
		}
	}
	return rules
}

```

Reply:
Done